base/bucket_gocb.go

```diff
@@ -987,6 +1135,68 @@ func (bucket CouchbaseBucketGoCB) WriteUpdate(k string, exp int, callback sgbuck
 
 }
 
+func (bucket CouchbaseBucketGoCB) WriteUpdateWithXattr(k string, xattr string, exp int, callback sgbucket.WriteUpdateWithXattrFunc) error {
+
+	bucket.singleOps <- struct{}{}
+	defer func() {
+		<-bucket.singleOps
+	}()
+
+	for {
+		var value []byte
+		var xattrValue []byte
+		var err error
+		var cas uint64
+
+		// Load the existing value.
+		gocbExpvars.Add("Update_GetWithXattr", 1)
+		cas, err = bucket.GetWithXattr(k, xattr, &value, &xattrValue)
+		if err != nil {
+			if !bucket.IsKeyNotFoundError(err) {
+				// Unexpected error, cancel writeupdate
+				LogTo("gocb", "Retrieval of existing doc failed during WriteUpdateWithXattr for key=%s, xattr=%s: %v", k, xattr, err)
+				return err
+			}
+			// Key not found - initialize cas and values
+			cas = 0
+			value = nil
+			xattrValue = nil
+		}
+
+		// Invoke callback to get updated value
+		updatedValue, updatedXattrValue, err := callback(value, xattrValue)
+		if err != nil {
+			LogTo("gocb", "Callback in WriteUpdateWithXattr returned error for key=%s, xattr=%s: %v", k, xattr, err)
+			return err
+		}
+
+		// TODO: Avoid unmarshalling the raw xattr value here
+		var xattrMap map[string]interface{}
+		err = json.Unmarshal(updatedXattrValue, &xattrMap)
+		if err != nil {
+			return err
+		}
+
+		// CAS-safe write of the new value and xattr
+		LogTo("gocb", "Calling WriteCasWithXattr for key: %s cas: %d, value: %v, xattr:%v", k, cas, updatedValue, xattrMap)
+		_, err = bucket.WriteCasWithXattr(k, xattr, exp, cas, updatedValue, xattrMap)
+		// ErrKeyExists is CAS failure, which we want to retry.  Other non-recoverable errors should cancel the
+		// WriteUpdate.
+		if err != nil && err != gocb.ErrKeyExists && !isRecoverableGoCBError(err) {
+			LogTo("gocb", "Update of new value during WriteUpdateWithXattr failed for key %s: %v", k, err)
+			return err
+		}
+
+		// If there was no error, we're done
+		if err == nil {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("Failed to update")
```

Comment:
is it even possible to get here?  if so, would it be possible to capture more detailed error information?